torch/csrc/jit/runtime/interpreter.cpp

```diff
@@ -1044,28 +1067,62 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
 
     // RecordFunction object associated with this frame
     std::unique_ptr<at::RecordFunction> record_function;
+
     // symbol table for a frame
     ShapeSymbolTable symbols2dims;
+
+    // call site source range
+    c10::optional<CallSiteInfo> call_info;
   };
 
   // saved-by-value stuff that can exist on the stack inside runInterpreter
   struct ActiveFrame {
     size_t pc;
     Instruction* instructions;
+    Node** instructions_source;
     IValue* constants;
     Operation* operators;
     Function** functions;
     std::function<void(std::vector<IValue>&)>* profile_functions;
     TypePtr* types;
 
-    ActiveFrame(const Frame& frame)
+    ActiveFrame(Frame& frame, Stack& stack)
         : pc(frame.pc),
           instructions(frame.function->instructions_.data()),
+          instructions_source(frame.function->instructions_source_.data()),
           constants(frame.function->constant_table_.data()),
           operators(frame.function->operator_table_.data()),
           functions(frame.function->function_table_.data()),
           profile_functions(frame.function->profile_function_table_.data()),
-          types(frame.function->type_table_.data()) {}
+          types(frame.function->type_table_.data()) {
+      // set the current call site info to the call site of the frame
+      // (e.g. function/method call, fork, etc)
+      tls_call_info_ = frame.call_info;
+
+      // if that's the first time we activate the frame, start
+      // the record function
+      if (frame.pc == 0) {
+        checkAndStartRecordFunction(frame, stack);
```

Comment:
it's weird that you put it into ActiveFrame constructor. Seems like `runGraphFunction` is a better place where we actually enter a new function. Or is it to make it work with "suspendible" tasks? Then I'm afraid it won't generate the nested events beyond the first level nesting anyway

Reply:
i moved it out of runGraphFunction because there are other cases when we'd want to call RecordFunction, e.g. when we do fork (vs in-script function/method call)

passing stack into ActiveFrame - that's purely for RecordFunction in cases when this is the first time we execute the frame,
and it guarantees we don't forget to do it,
originally i just called checkAndStartRecordFunction after the several use cases of ActiveFrame, we could return to that